sample_files/problem_setting/examples/identical_checker.cpp

```diff
@@ -0,0 +1,132 @@
+#include <algorithm>
+#include <cstdio>
+#include <cstdlib>
+#include <fstream>
+#include <numeric>
+#include <regex.h>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace CheckerCodes {
+int AC = 0;
+int WA = 1;
+int PE = 2;
+int PARTIAL = 7;
+} // namespace CheckerCodes
+
+namespace regex_helpers {
+regex_t compile(const char *pattern) {
+  regex_t re;
+  if (regcomp(&re, pattern, REG_EXTENDED | REG_NOSUB) != 0) {
+    throw std::runtime_error("Pattern failed to compile.");
+  }
+  return re;
+}
+bool match(regex_t re, const std::string &text) {
+  return regexec(&re, text.c_str(), 0, NULL, 0) == 0;
+}
+} // namespace regex_helpers
+
+void errorHook();
+void exitWA() {
+  errorHook();
+  std::exit(CheckerCodes::WA);
+}
+void exitPE() {
+  errorHook();
+  std::exit(CheckerCodes::PE);
+}
+void assertWA(bool condition) {
+  if (!condition) {
+    exitWA();
+  }
+}
+void assertPE(bool condition) {
+  if (!condition) {
+    exitPE();
+  }
+}
+void readSpace() { assertPE(getchar() == ' '); }
+void readNewLine() { assertPE(getchar() == '\n'); }
+void readEOF() { assertPE(getchar() == EOF); }
+std::string readToken(char min_char = 0, char max_char = 127) {
+  static constexpr size_t MAX_TOKEN_SIZE = 1e7;
+  std::string token;
+  int c = getchar();
+  assertPE(!isspace(c));
+  while (!isspace(c) && c != EOF) {
+    assertWA(token.size() < MAX_TOKEN_SIZE);
+    assertWA(min_char <= c && c <= max_char);
+    token.push_back(char(c));
+    c = getchar();
+  }
+  ungetc(c, stdin);
+  return token;
+}
+long long readInt(long long lo, long long hi) {
+  static regex_t re = regex_helpers::compile("^(0|-?[1-9][0-9]*)$");
+  std::string token = readToken();
+  assertWA(regex_helpers::match(re, token));
+
+  long long parsedInt;
+  try {
+    parsedInt = stoll(token);
+  } catch (const std::invalid_argument &) {
+    exitWA();
+  } catch (const std::out_of_range &) {
+    exitWA();
+  }
+  assertWA(lo <= parsedInt && parsedInt <= hi);
+  return parsedInt;
+}
+template <typename T>
+std::vector<T> readIntArray(size_t N, long long lo, long long hi) {
```

Comment:
Should probably add as a comment that this will need modification if you want to read `unsigned long long`.

Reply:
Added to docs. I think fixing this function to handle uint64 is a pain... The best option that comes to mind is having `readInt` have a template parameter for the return type. Inferring template parameters from the arguments, like doing `readInt(T lo, T hi)` doesn't work because the compiler shits itself if you try `readInt(0, 1e5)` or `readInt(1, 4U)` because the types are different.

Hard-requiring uint64 is rare, so I think this limitation is ok, since the alternatives sacrifice ergonomics greatly.

Reply:
Don't you need to specify a template type anyway since `T` doesn't appear in the argument list at all? So the problem you are trying to address with type inference doesn't actually exist...

Reply:
Yes, `readIntArray` forces specifying the template parameter and this is intentional - it's impossible to ergonomically infer it from the arguments, and having the return value always be long longs is not great, unlike with `readInt` where immediate downcasting is possible.

Reply:
Then, just declare `lo` and `hi` as `T` to avoid pointlessly adding limitations.

Reply:
For reasons discussed before this isn't possible. Specifically, `readIntArray(N, 1, 1e9)` no longer works in that case.

Reply:
`readIntArray(N, 1, 1e9)` never works because you always need to specify a return type due to the template being on the return value only. So you are optimizing for a case that doesn't exist.

If anything, using `T` in the parameter list will allow `readIntArray(N, 1, 100);` to work.

Reply:
Right, so to make `readIntArray(N, 1, 1e9)` work, we force passing the template parameter. Then `readIntArray<int>(N, 1, 1e9)` works.

Reply:
The proper way to do this in C++ if you care about type deduction would be something like:

```cpp
template<typename T>
T readInteger(T lo, T hi) {
   // ...
}

constexpr auto readInt = &readInteger<int>;
// more aliases as needed

template<typename T>
std::vector<T> readIntArray(size_t N, T lo, T hi) {
    // ...
}
```

You know what, this is always going to be terrible and I am done trying to police this into something sane.